hudi-utilities/src/main/java/org/apache/hudi/utilities/HoodieMetadataSync.java

```diff
@@ -0,0 +1,529 @@
+          fileIdToBaseInstantTimeToFileName.computeIfAbsent(fileId, s -> new HashMap<>());
+          fileIdToBaseInstantTimeToFileName.get(fileId).put(baseInstantTime, fileName);
+        });
+
+        List<String> deletePathPatternsFromTarget = new ArrayList<>();
+        fileIdsToDeletedFiles.forEach((k, v) -> {
+          String fileId = k;
+          List<Pair<String, String>> fileNameAndBaseInstantTime = v;
+          fileNameAndBaseInstantTime.forEach(pair -> {
+            if (fileIdToBaseInstantTimeToFileName.get(fileId) != null && fileIdToBaseInstantTimeToFileName.get(fileId).containsKey(pair.getValue())) {
+              //System.out.println("Source file id " + k + ", base instant time " + pair.getRight() + ", file name " + pair.getLeft() + " => target table file Name " + fileIdToBaseInstantTimeToFileName.get(fileId).get(pair.getRight()));
+              deletePathPatternsFromTarget.add(fileIdToBaseInstantTimeToFileName.get(fileId).get(pair.getRight()));
+            } else {
+              //System.out.println("Source file id " + k + ", base instant time " + pair.getRight() + ", file name " + pair.getLeft() + " => No matching file found in target table XXXXXXX ");
+            }
+          });
+        });
+        if (cleanPartitionMetadata.getDeletePathPatterns().size() == deletePathPatternsFromTarget.size()) {
+          cleanPartitionMetadata.setDeletePathPatterns(deletePathPatternsFromTarget);
+        } else {
+          throw new HoodieException("Failed to find matching file from target table");
+        }
+      });
+
+      return originalCleanMetadata;
+    } finally {
+      if (tableMetadata != null) {
+        try {
+          tableMetadata.close();
+        } catch (Exception e) {
+          throw new HoodieException("Failed to close Table Metadata", e);
+        }
+      }
+    }
+  }
+
+  private HoodieTableFileSystemView getFSV(HoodieWriteConfig config, FileSystemViewStorageType storageType, HoodieSparkEngineContext context, HoodieTableMetaClient metaClient) {
+    FileSystemViewStorageConfig viewStorageConfig = FileSystemViewStorageConfig.newBuilder().fromProperties(config.getProps())
+        .withStorageType(storageType).build();
+    return (HoodieTableFileSystemView) FileSystemViewManager
+        .createViewManager(context, config.getMetadataConfig(), viewStorageConfig, config.getCommonConfig(),
+            (SerializableFunctionUnchecked<HoodieTableMetaClient, HoodieTableMetadata>) v1 ->
+                HoodieTableMetadata.create(context, config.getMetadataConfig(), metaClient.getBasePathV2().toString(), true))
+        .getFileSystemView(config.getBasePath());
+  }
+
+  private List<HoodieInstant> getInstantsToSync(String sourceIdentifier, HoodieTableMetaClient targetTableMetaClient, HoodieTableMetaClient sourceTableMetaClient) throws IOException {
+    Option<HoodieInstant> lastInstant = targetTableMetaClient.reloadActiveTimeline().getWriteTimeline().filterCompletedInstants().lastInstant();
```

Comment:
here we are looking at only the last instant. 
So, from locking standpoint, do we know we are going to fully take a table level lock before we do the entire metadata sync. 
and we always encode all table checkpoints in every commit. 

and so, its safe to read from latest?